11 integrate the GitHub action ci with cat-github-actions

sooyean-hoo opened the pull request:
- Ref: 
> https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/issues/11 
> https://github.com/puppetlabs/puppetlabs-servicenow_cmdb_integration/issues/72 #72 


TLDR; 

The line which draw the conclusion that the acceptance is only for Ubuntu is @ here.  https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/blob/a4c76f6b718ac9b9ee824324a904ece15b04432d/spec/support/acceptance/install_pe.sh#L8

The file install_pe.sh is used in all variant, hence all tests will failed for now, esp. even O/S Ubuntu 16.04 is not longer supported.

The task originally was to just to update the tests to bring it to date. However now it turned out to be a bigger task: writing test to work with all the missing O/S. about 10 of them. This is way more complicated, as there is also need to add more scenario and cater/reimplement for all old functions and original testing goals in the new O/S.

Now. 
So now only *Oracle Linux 7*, *RedHat Linux 7 and 8*, *Ubuntu 18.04, 20.04 and 22.04* are
 part of the testing Matrix. The exclusions and the reason for it are listed below.
 As an overview, there are a lot difficult in running the mock-up `ServiceNow server`.
 It is still run as a docker, but for Ubuntu, they are hosted in the `PE Primary server`.
 For others, they are hosted in the Runner. This is because I only managed to get docker to
 play nice in Ubuntu. And yes the Runner is also of Ubuntu.
 There are playful manipulation of port forwardings and /etc/hosts used to standardise 
 the access to both the `PE primary server `and the` mock-up ServiceNow server`.
 https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/actions/runs/10669025799

![image](https://github.com/user-attachments/assets/52b255b5-4386-44af-b7d6-747631229727)

sooyean-hoo committed:
=== Testing against Oracle Linux 7 & 8, RedHat Linux 7 and 8, Ubuntu
18.04, 20.04 and 22.04 ===
add anew parameter snow_uri_erb which can optionally specified the URI
to use for connecting to the ServiceNow server. This allows for more
flexible configuration
PDK Validated
Add an option to all the request level as a parameter... also use
binding for erb.
Added in the debug options so that the test can be carried out.

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Added in CI.
Update to allow for run
pdk validated
PDK Validated
CI updated again
PDK Validated again
Disabled acceptances
Ci change to puppetlabs's
==vagrant-best_bet==              Update provision to change to docker
instead of vmpooler.. which seems
to keep failing and use ubuntu 18.04
Update the changes of node-manager

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Add in github_changelog_generator
USE Docker... Use InstallPKG.... Move Service Now into Docker ..Skip
provision in Prep
....  Disabled
Spec and enabled Redhat 8 for
Accept..    New Docker workflow
Update it and adjusting the puppetversion.. as the
old version DOcker Images is deprecated,  add Setup_Servicenow_instance.
FFI issue . Add OLD CI for reference, with update runners

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Use back vagrant version
Upgrade the vagrant
Add ininstallpecommands

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Remove a centos Vagrant provision
Skip some setup sub parts.
Add in bundle install
Testing ssh @ 2222
Move the old files.. to a deprecated folder.. They are not working at
all. And functionaries are replaced by ci.yaml
More info

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
More vagrant info
more vagrant tweaking

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Vagrant exploring
Testing new Vagrant access
Firm up Vagrant access with own keys
Install all the agents

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Submodule added

Now Run Tests more roles
Test run with flags
Reshape the Inventory
==Working with_system Test_Completed with proposed_CI==
Back to Docker
More Docker settings
Docker adjustments
Correct the bash scripts
Switch to a new design, and update the scripts

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Set up preinstallpecommands
Did the preinstall ops
ordering of the tests
swap task place
Inventory updates

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Adjust for password injection adjustment for vagrant VMs as primary
server
Add missing load lib
Locate script ./spec/support/acceptance/install_pe.sh
Adjusted the path
Adjust for vagrant more
MOre adjustment to the inv for the acceptance test.
Add in the tests of the ports
ping_NC_Test fail
Connectivity Checks Before  Running the rest
master
ServiceNow code
Run with checks again
Intro vhelper.rb to streamline codings
Update the bash part
Old CI update for benchmark

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Lint friendly

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Change back to original cannot get to work either way
Try a new docker style
To vagrant again.. since docker cannot get pe to work... as it cannot
start cron service in docker.
Hard code localhost for now. for vagrant now
Port forwarding...
add in error id set timeout...
Longer timeout.
get connected.. with some delay
Fix a proxy error
Try to fix host verification
Remove the port also from the uri
show hostkey
prevent repeated entry in knownhost.
ssh know data
Set master dns
Ensure we are using ${pehostnameinservicenow} as the DNS of master..
which is example.puppet.com
Change to use host for connection
Added in         ssh-keyscan -t rsa ${pehostnameinservicenow}   >>
$HOME/.ssh/known_hosts ;

redo the known hosts.
HostCheck
Find package modules
Ownself install Package Modules
Update vhelper.rb scp param
Skip host verification
Revert some inventory settings
Add back in the port
Clean up the port suffix
PDK Validated
Update for using Certname instead
mod test a nit
Tear down error solved
Complete the help command
Complete the help
Run the acceptant test
Adding the option of InstallOpenSource
Change to ospversion
Revert to opensourcepuppetversion
Keep the OK  version so it is better
OpenSource Puppet server
==FIX PE Install====
Exposing more Rake tasks and comforting to the Litmus standards
Switch to || true where possible
Added in the proposed CI
PDK Validated
CI proposed
UPdate the PE Version
UPdate the pe and gems pdk validated
Add spaces
New Proposed with TearDown
MOre setup
See env
Use run chain
Prep Install and prep
use Quotes
Show env -test
PDK validated
Remove Gem updates
New approach for shell
Running with only steps slots from   module_acceptance.yml

Remove some extra lines
Use Open3 for acceptance tests
PDK validated
Add postcommand
Better clean up and  handler or error code
Make the error code passed correctly...
Fix a spec error
New matrix
Correct the return the right exit code
Update vhelper.rb remove spaces
Testing Gemfile upgrade
Revert Back to the Gemfile
Testing for cat github action
New Mapping for the provision
Testing module_acceptance_custom
Add in vscode extension
Update on an error on  exist tatus
Added in Provision check for various OS
better provision translation
Update with new OS and naming convention
Update the mapping in provision.
Exclude the wrong ones
Check for empty inventory file
So add Cannot find  generic version for OracleLinux-7, SLES-12,
Scientific-7,  so  Skipped them in tests.  So use some other versions.
ADDED matrix_from_metadata function
Update for sles12sp4
Add in oraclelinux 8 and Scientific 8
New      matrix_from_metadata

Show Original
New matrix_from_metadata
Remove some of obsolete Workflow
Got the right Vagrant folder
Cleaning up Installer and less verbose
Ass in failsafe PE install
servicenow install updated
Check servicenow service setup and catch failure
Try some stuff for RHEL
Added support for SLES
Enabled Extra Repo
Testonly what was needed
Move setupServiceNowServer fn up
SLES resolved???
Final check reorder
SLES try again
More failsafe for SLES
Added in    convenience script
Added dockerconveniencescript
For force a local one
Try for better docker
Install CURL only if needed
Added Creating FailSafe ServiceNow Server
PDK Validated
some explain on exception addn ignore.. cannot
Restore back the ploygotism
Do the service now on the runner as a failsafe...
PDK Validated

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Update metatdata for Centos and
Add more failsafe into install PE.

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Docker add more fixes.
Fix the docker issue and added a test.
Safer
Failsafe complete with testing.
FOr dockers
Set up correct path to puppet
Add in checks for ServiceNow node
clear away extra O/S ensure puppet token is generated
Try to get the system to use the runner docker for service now

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Another way to do the servicenow server
Allow for a better determination for service now   setupServiceNowServer
Fix some obvious mistake
More specific with the servicenow server
Adjust output... echo provision
More Formating of the output.
Cat Me the file
Do the service now set up early
Better setup
Re arrange back  the setup order
Fix a nil error adjust port fwd
Add more debugging data for servicenow setup.... Localhost may not work
for Non Ubuntu machines
Reverse port serving for service now ... on testing of non-ubuntu
machines
Setting in puppet into path to make it for ALL OSes
A better way specify the PE installation and Servicenow Checking
Fix install_pe

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
===RHEL WOrks==
Add export to path

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
dnf for REDHAT* and more RHEL stuff and OracleLinux
Config Hostsname on the Runner
PDK Validated
Stop the blocking of errors

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Added Required tests
Checks are spelt out...
Clean up my small setup tests.
Filter more O/S , Add firewall for PE and more hostname
More checks for RHEL to force force for installation and filter away
working Ubuntus
Shift to x-64
No Bolt as gems
generic-x64 to generic
Fixing a silly mistake
Put back bolt gem
change bacj metadata
Revert a lot
Switch to new matrix_from_metadata
Support for RHEL
==Locale fixed For RHEL installation ====

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
added  checkSELINUX disableSELINUX    restartCompilersReplicaServices
Mod SSHD and Force the use of bolt binary
Using an alternative method to bypass bolt command to use the binary..
so no need to  gem uninstall
Set to use the correct bolt
SSHd config and  Folder permission ensured.... add in
-oIdentitiesOnly=yes to avoid prompt
chmod sudo first
Install git first
Pushing Installatoin for Oracle Linux 7
Split the provisioning to into for easier management... cos part 2 is
always missing
Set in BOLT_PROJECT=

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Split more parts..
redacted all passwords

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Force bolt commands
Set less error msg for setting permission
Show the error code... Need to filter away or wait for run which result
in 'Run of Puppet configuration client already in progress; skipping'
Split install_module__task

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Fixing a pipe mistake
Error Code 1 is Run of Puppet configuration client already in progress;
skipping, WE SHOULD WAIT 120secs for 3 times AND TRY AGAIN.
Add a new check whichpuppet
Remove RedHat 9 too much work... Save for other day
Based off
https://www.puppet.com/docs/pe/2023.6/running_tasks_from_the_command_line#puppet-task-run-options
     Seems to having problem as it cannot handle array :

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Revert to how they call the puppet command
MOre checkings
which puppet is done.
Put puppet in PATH
Part 2 Add  VALENTEHOME to allow for local debugging and add better
which
puppet checks and fix
Fixing Missing Puppet Command
Nex adjustment
more log
Git install and extra configuration for more verbose and fix the port
fwd error
Intro coy.servicenow as servicenow server DNS
PDK Validated
Shifted to coy.servicenow

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
coy.servicenow for nonubuntu
More log to reveal more
Fix more serviceNOW dns error
For local fwd also.
Revert to old way to passing the servernow conn over.. all tested now.

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Setup the Configure File and ensure it is writable. To prevent Error in
RHEL7 and Oracle Linux
mkdir -p /etc/puppetlabs/puppet

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Revert to local host for servicenow.
Ensure RW of /etc/puppetlabs/puppet and files
add  'export LC_ALL="en_US.UTF-8" ;' to  $HOME/.bashrc

===CLEAN UP for deployment ==
Extend the length for the run
Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
creat the file first
Handled FALSE NEGATIVE: 'Run of Puppet configuration client already in
progress; skipping', WE SHOULD WAIT 120secs for 3 times AND TRY AGAIN.
Adding # POLYGLOT tags for the function monoglot

Rename ployglot version to ./vhelper.sh.rb
Adding function monoglot
Add in the monoglot versions and ensure the reference is update for
monoglot versions.
Updated Github actions.
PDK validate and update the monoglot to allow for permission setting
Clean up all the extra development resources

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Merge branch
'release/11-refactor-extend-and-develop-to-integrate-the-github-action-ci-with-httpsgithubcompuppetlabscat-github-actionstreemaingithub'
into main

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>
Addin OracleLinux 6   Add in the OracleLinux for production...
Add in OracleLinux8
Update metadata

Signed-off-by: Hoo Sooyean (何書淵) <Sooyean.hoo@perforce.com>

gregohardy commented:
Hi @sooyean-hoo. Great stuff. Dont forget to squash your commit so we can review it.

sooyean-hoo commented:
@gregohardy I squashed it before... to  a few commits actually... Anyway with the recent update, I squashed it further to only 1 commit... 

Along with the squash, I add a new O/S to the matrix. I cannot explain why I can get OracleLinux 7 and RHEL-8 working, but not OracleLinux 8....... 

I was also puzzled why I wrote... https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/blob/386d71745ace9395d91a7d09ac5bee95ec393a97/.github/workflows/ci_proposed.yml#L78

```
# \"OracleLinux-8\", = ?????
```

Hence I did a review on OracleLinux 8. It turns out that `matrix_from_metadata_v2`  @ https://github.com/puppetlabs/puppet_litmus/blob/v0.36.1/exe/matrix_from_metadata_v2 (version 0.36.1) will reject OracleLinux 8 as it cannot find a match.

Hence I cannot translate it into a vagrant VM. To which I need to do some 'magic' to include it.

# **Long Story short, I squashed the commits into 1 and extended the number of O/Ses in the testing matrix from 6 to 7, with the new O/S as `Oracle Linux 8` and the original 6 O/Ses as  `OracleLinux 7`, `Ubuntu 18.04, 20.04 22.04` and `RedHat 7 & 8` . **

As the time of this comment, the tests are still running. Hence I am not sure how good or bad the results are. Sometimes it needs a bit of nudge aka `re-run the failed one` to work. I did test before in another branch. It was successful for the `OracleLinux 8`. See: https://github.com/sooyean-hoo/puppetlabs-servicenow_cmdb_integration/actions/runs/10679611998/job/29599273899

actowery committed:
Merge branch 'main' into dev_11-refactor-extend-and-develop-to-integrate-the-github-action-ci-with-httpsgithubcompuppetlabscat-github-actionstreemaingithub
